### PR TITLE
Add a class for accumulating signals

### PR DESCRIPTION
**Pull request by cmcantalupo**
Performs sums or averages of sampled signals and is a helper class for the Reporter or an agent that wants to make measurements of regions or epochs.

### CODE REVIEW

File: test/AccumulatorTest.cpp [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+
+TEST_F(AccumulatorTest, empty)
+{
+    EXPECT_EQ(0.0, m_accum_sum->sample_all());
+    EXPECT_EQ(0.0, m_accum_sum->sample_last());
+    EXPECT_EQ(0.0, m_accum_avg->sample_all());
+    EXPECT_EQ(0.0, m_accum_avg->sample_last());
+}
+
+TEST_F(AccumulatorTest, sum_ones)
```
Comment (cmcantalupo) [inline test/AccumulatorTest.cpp:63]: Add test links to test/Makefile.mk

File: src/Accumulator.hpp [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+#define ACCUMULATOR_HPP_INCLUDE
+
+#include <memory>
+
+namespace geopm
+{
+    class SumAccumulator
+    {
+        public:
+            static std::unique_ptr<SumAccumulator> make_unique(void);
```
Comment (cmcantalupo) [inline src/Accumulator.hpp:74]: Add doxygen

File: src/Accumulator.hpp [hunk 3]
```diff
@@ -37,30 +37,141 @@
...
+            /// by an entry() and exit() call.
+            ///
+            /// @return Sum of all values passed to update() during
+            ///         last interval.
+            virtual double interval_total(void) const = 0;
         protected:
             SumAccumulator() = default;
     };
 
     class AvgAccumulator
```
Comment (cmcantalupo) [inline src/Accumulator.hpp:143]: Add dox

File: src/Accumulator.hpp [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+ */
+
+#ifndef ACCUMULATOR_HPP_INCLUDE
+#define ACCUMULATOR_HPP_INCLUDE
+
+#include <memory>
+
+namespace geopm
+{
+    /// @brief Class to track the accumulation of a signal.
```
Comment (dianarg) [inline src/Accumulator.hpp]: maybe a sentence here to motivate the "condition" to be tracked, e.g. enables per-region or per-hint accumulation of signals

File: src/Accumulator.hpp [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+    /// the application, the epoch events, or the hint signal.  The
+    /// user only calls the update() method when the condition is true
+    /// (e.g. the application is within the tracked region).  The
+    /// enter() and exit() API's are used to track values for the last
+    /// occurrence of the condition being true.  It is expected (though
+    /// not enforced) that one call to enter() proceeds each call to
+    /// exit(), and these are used to update the values returned by
+    /// interval_total().
+    ///
+    /// The SumAccumulator is used to accumulate a signal that is
```
Comment (dianarg) [inline src/Accumulator.hpp:53]: I would move this paragraph up after the first

File: src/Accumulator.hpp [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+            /// Used to report on the total accumulated sum of all of
+            /// the updates since the construction of the object.
+            ///
+            /// @return Sum of all values passed to update()
+            virtual double total(void) const = 0;
+            /// @brief Increase of tracked signal over last interval.
+            ///
+            /// Get the increase in the signal while the condition is
+            /// true over the last interval.  An interval is defined
+            /// by an entry() and exit() call.
```
Comment (dianarg) [inline src/Accumulator.hpp]: enter()

File: src/Accumulator.hpp [hunk 7]
```diff
@@ -0,0 +1,214 @@
...
+        public:
+            SumAccumulatorImp();
+            virtual ~SumAccumulatorImp() = default;
+            void update(double delta_signal) override;
+            void enter(void) override;
+            void exit(void) override;
+            double total(void) const override;
+            double interval_total(void) const override;
+        private:
+            double m_total; // Sum of all delta values
```
Comment (dianarg) [inline src/Accumulator.hpp]: doxygen style for these, or remove?

**Review by dianarg — APPROVED** [review]

**Pull request merged**